compiler/src/iree/compiler/Dialect/Stream/IR/StreamOpFolders.cpp

```diff
@@ -313,6 +313,38 @@ struct ElideUnusedOp : public OpRewritePattern<Op> {
   }
 };
 
+// Clones ops that prefer to be cloned directly.
+// This prevents us from splatting out a value and then cloning that (keeping
+// the memory live/etc) instead of just splatting it again on-demand.
+//
+// Example:
+//  %0 = stream.async.splat %c123_i32
+//  %1 = stream.async.clone %0
+// ->
+//  %1 = stream.async.splat %c123_i32
+template <typename Op>
+struct PropagateClonableOps : public OpRewritePattern<Op> {
+  using OpRewritePattern<Op>::OpRewritePattern;
+  LogicalResult matchAndRewrite(Op cloneOp,
+                                PatternRewriter &rewriter) const override {
+    if (cloneOp.use_empty()) return failure();
+    auto sourceOp =
+        cloneOp.getSource()
+            .template getDefiningOp<IREE::Stream::StreamableOpInterface>();
+    if (!sourceOp || !sourceOp.preferCloneToConsumers()) return failure();
+    for (auto &use :
+         llvm::make_early_inc_range(cloneOp.getResult().getUses())) {
+      rewriter.setInsertionPoint(use.getOwner());
+      auto clonedOp = rewriter.clone(*sourceOp);
+      use.set(clonedOp->getResult(0));
```

Comment:
I think Pattern Rewriter has some quirks where not using pattern rewriter sometimes (even to update operands) sometimes blows things up.... Just FYI (and I might be totally wrong). Again no request here.